components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1449 @@
+}
+
+fn round_frac(frac: u32, fsp: u8) -> (bool, u32) {
+    debug_assert!(frac < 100_000_000 && fsp < 7);
+    let fsp = usize::from(fsp);
+    let width = if frac >= 1_000_000 { 7 } else { 6 };
+    let mask = TEN_POW[width - fsp - 1];
+    let result = (frac / mask + 5) / 10 * mask * if width == 6 { 10 } else { 1 };
+    (result >= 1_000_000, result)
+}
+
+bitfield! {
+    #[derive(Clone, Copy, Default)]
+    pub struct Time(u64);
+
+    u32;
+    #[inline]
+    get_year, set_year: 63, 50;
+    #[inline]
+    get_month, set_month: 49, 46;
+    #[inline]
+    get_day, set_day: 45, 41;
+    #[inline]
+    get_hour, set_hour: 40, 36;
+    #[inline]
+    get_minute, set_minute: 35, 30;
+    #[inline]
+    get_second,set_second: 29, 24;
+    #[inline]
+    get_micro, set_micro: 23, 4;
+
+    // `fsp_tt` format:
+    // | fsp: 3 bits | type: 1 bit |
+    // When `fsp` is valid (in range [0, 6]):
+    // 1. `type` bit 0 represent `DateTime`
+    // 2. `type` bit 1 represent `TimeStamp`
+    //
+    // Since `Date` does not require `fsp`, we could use `fsp` == 0b111 to represent it.
+    #[inline]
+    u8, get_fsp_tt, set_fsp_tt: 3, 0;
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum TimeType {
+    Date,
+    DateTime,
+    TimeStamp,
+}
+
+impl TryFrom<FieldTypeTp> for TimeType {
+    type Error = codec::Error;
+    fn try_from(time_type: FieldTypeTp) -> Result<TimeType> {
+        Ok(match time_type {
+            FieldTypeTp::Date => TimeType::Date,
+            FieldTypeTp::DateTime => TimeType::DateTime,
+            FieldTypeTp::Timestamp => TimeType::TimeStamp,
+            _ => return Err(box_err!("Time does not support field type {}", time_type)),
+        })
+    }
+}
+
+// The common set of methods for `date/time`
+impl Time {
+    /// Returns the hour number from 0 to 23.
+    #[inline]
+    pub fn hour(self) -> u32 {
+        self.get_hour()
+    }
+
+    /// Returns the minute number from 0 to 59.
+    #[inline]
+    pub fn minute(self) -> u32 {
+        self.get_minute()
+    }
+
+    /// Returns the second number from 0 to 59.
+    #[inline]
+    pub fn second(self) -> u32 {
+        self.get_second()
+    }
+
+    /// Returns the number of microseconds since the whole second.
+    pub fn micro(self) -> u32 {
+        self.get_micro()
+    }
+
+    /// Returns the year number
+    pub fn year(self) -> u32 {
+        self.get_year()
+    }
+
+    /// Returns the month number
+    pub fn month(self) -> u32 {
+        self.get_month()
+    }
+
+    /// Returns the day number
+    pub fn day(self) -> u32 {
+        self.get_day()
+    }
+}
+
+mod parser {
+    use super::*;
+    fn bytes_to_u32(input: &[u8]) -> Option<u32> {
+        input.iter().try_fold(0u32, |acc, d| {
+            d.is_ascii_digit().as_option()?;
+            acc.checked_mul(10)
+                .and_then(|t| t.checked_add(u32::from(d - b'0')))
+        })
+    }
+
+    fn digit1(input: &[u8]) -> Option<(&[u8], &[u8])> {
```

Comment:
Add doc for what the return value is (`(reset, digit)`), or define a struct as the return type. Or the maintainer will easily to make mistake here

Reply:
OK.

Reply:
@iosmanthus cc
![image](https://user-images.githubusercontent.com/21231345/66695151-8ccc0d80-ecf0-11e9-9886-3a873e6900d2.png)
Why there is no doc here.
BTW, I think return a struct is better. Because the type of this tuple's two part is the same. So It is very easy to make mistake here.

Reply:
Perfer make `digit1` be inner function of `split_components`, because `digit1`'s param is `[u8]`, so maintainer may make mistake of using it.